tockloader-lib/src/tabs/tab.rs

```diff
@@ -22,36 +22,48 @@ impl Tab {
         let mut metadata = None;
         let mut tbf_files = Vec::new();
 
-        let mut archive = Archive::new(File::open(path).unwrap());
-        for file in archive.entries().unwrap() {
-            let mut file = file.unwrap();
-            let path = file.path().unwrap();
-            let file_name = path.file_name().unwrap().to_str().unwrap().to_owned();
+        let file = File::open(path).map_err(TockloaderError::UnusableTab)?;
+        let mut archive = Archive::new(file);
+        for file in archive
+            .entries()
+            .map_err(TockloaderError::UnusableTab)?
+        {
+            match file {
+                Ok(mut file) => {
+                    let path = file.path().map_err(TockloaderError::UnusableTab)?;
+                    match path.file_name() {
+                        Some(file_name) => {
+                            let file_name = file_name.to_str().unwrap_or("").to_owned();
+                            if file_name == "metadata.toml" {
+                                let mut buf = String::new();
+                                file.read_to_string(&mut buf)
+                                    .map_err(TockloaderError::UnusableTab)?;
+                                metadata = Some(Metadata::new(buf)?);
+                            } else if file_name.ends_with(".tbf") {
+                                let mut data = Vec::new();
 
-            if file_name == "metadata.toml" {
-                let mut buf = String::new();
-                file.read_to_string(&mut buf).unwrap();
-
-                metadata = Some(Metadata::new(buf).unwrap());
-            } else if file_name.ends_with(".tbf") {
-                let mut data = Vec::new();
-                file.read_to_end(&mut data).unwrap();
-
-                tbf_files.push(TbfFile {
-                    filename: file_name.to_string(),
-                    data,
-                });
+                                file.read_to_end(&mut data)
+                                    .map_err(TockloaderError::UnusableTab)?;
+                                tbf_files.push(TbfFile {
+                                    filename: file_name.to_string(),
+                                    data,
+                                });
+                            }
+                        }
+                        None => continue,
+                    }
+                }
+                Err(e) => return Err(TockloaderError::UnusableTab(e)),
             }
         }
 
-        if metadata.is_none() {
-            panic!("No metadata.toml found in tab");
+        match metadata {
```

Comment:
map_err

Reply:
metadata is an Option so I can't use map_err

Reply:
ok_or

Reply:
Actually, It's fine this way. I just notice you make it go from metadata to a TAB. It's better as-is actually